File: 23-Graph/python/05_detect_a_cycle_in_an_undirected_graph_using_dfs.py
from typing import List

class Solution:
    def isCycle(self, V: int, adj: List[List[int]]) -> bool:
        visited = [False] * V

        def dfs_check(node: int, parent: int) -> bool:
            visited[node] = True
            for neighbor in adj[node]:
                if not visited[neighbor]:
                    if dfs_check(neighbor, node):
                        return True
                elif neighbor != parent:
                    return True
            return False

        for i in range(V):
            if not visited[i]:
                if dfs_check(i, -1):
                    return True
        return False

File: 23-Graph/cpp/05-detect-a-cycle-in-an-undirected-graph-using-dfs.cpp
#include <vector>

using namespace std;

class Solution {
private:
    bool dfsCheck(int node, int parent, vector<int> adj[], vector<bool>& visited) {
        visited[node] = true;
        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                if (dfsCheck(neighbor, node, adj, visited)) {
                    return true;
                }
            } else if (neighbor != parent) {
                return true;
            }
        }
        return false;
    }

public:
    bool isCycle(int V, vector<int> adj[]) {
        vector<bool> visited(V, false);
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                if (dfsCheck(i, -1, adj, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
};
